dockstore-webservice/src/main/java/io/dockstore/webservice/resources/EntryResource.java

```diff
@@ -92,6 +123,103 @@ public Entry updateAliases(@ApiParam(hidden = true) @Auth User user,
+            @ApiParam(value = "The id of the entry to add a topic to.", required = true) @PathParam("id") Long id,
+            @ApiParam(value = "The id of the category to add a topic to, defaults to Automatic Tool and Workflow Threads (6).", defaultValue = "6", allowableValues = "6,9") @QueryParam("categoryId") Integer categoryId,
+            @ApiParam(value = "This is here to appease Swagger. It requires PUT methods to have a body, even if it is empty. Please leave it empty.") String emptyBody) {
+        return createAndSetDiscourseTopic(id, categoryId);
+    }
+
+    /**
+     * For a given entry, create a Discourse thread if applicable and set in database
+     * @param id entry id
+     * @param categoryId category id (6 for automatic tools and workflows category)
+     * @return Entry with discourse ID set
+     */
+    public Entry createAndSetDiscourseTopic(Long id, Integer categoryId) {
+        Entry entry = this.toolDAO.getGenericEntryById(id);
+
+        if (entry == null || !entry.getIsPublished()) {
+            throw new CustomWebApplicationException("Entry " + id + " does not exist or is not published.", HttpStatus.SC_NOT_FOUND);
+        }
+
+        if (entry.getTopicId() != null) {
+            throw new CustomWebApplicationException("Entry " + id + " already has an associated Discourse topic.", HttpStatus.SC_BAD_REQUEST);
+        }
+
+        // Verify and set category
+        Integer category = defaultDiscourseCategoryId;
+        if (categoryId != null) {
+            if (categoryId.equals(testDiscourseCategoryId) || categoryId.equals(defaultDiscourseCategoryId)) {
+                category = categoryId;
+            } else {
+                throw new CustomWebApplicationException("Category " + categoryId + " is not a valid category.", HttpStatus.SC_BAD_REQUEST);
+            }
+        }
+
+        // Create title and link to entry
+        String entryLink = "https://dockstore.org/";
+        String title;
+        if (entry instanceof Workflow) {
+            title = ((Workflow)(entry)).getWorkflowPath();
+            entryLink += "workflows/";
+        } else {
+            title = ((Tool)(entry)).getToolPath();
+            entryLink += "tools/";
+        }
+
+        entryLink += title;
+
+        // Create description
+        String description = "";
+        if (entry.getDescription() != null) {
+            description = entry.getDescription() != null ? entry.getDescription().substring(0, Math.min(entry.getDescription().length(), maxDescriptionLength)) : "";
+        }
+
+        description += "\n<hr>\n<small>This is a companion discussion topic for the original entry at <a href='" + entryLink + "'>" + title + "</a></small>\n";
+
+        // Check that discourse is reachable
+        boolean isReachable;
+        HttpURLConnection connection = null;
+        try {
+            URL url = new URL(discourseUrl);
+            connection = (HttpURLConnection)url.openConnection();
```

Comment:
try with resources

Reply:
HttpURLConnection does not implement autocloseable so I can't.